### PR TITLE
Starter on #6

### PR DESCRIPTION
**Pull request by Robinlovelace**

### CODE REVIEW

**Comment by Robinlovelace** [comment]
I'm out of my depth here so any feedback v. welcome @dabreegster.

All the errors!

Starting with...

```
 Compiling zonebuilder v0.1.0 (/mnt/57982e2a-2874-4246-a6fe-115c199bc6bd/orgs/zonebuilders/zonebuilder-rust)
error: expected one of `::`, `:`, or `|`, found `(`
  --> src/lib.rs:26:22
```

<details>

```
 Compiling zonebuilder v0.1.0 (/mnt/57982e2a-2874-4246-a6fe-115c199bc6bd/orgs/zonebuilders/zonebuilder-rust)
error: expected one of `::`, `:`, or `|`, found `(`
  --> src/lib.rs:26:22
   |
26 |     ..params::default()
   |                      ^ expected one of `::`, `:`, or `|`

error: expected one of `::`, `:`, or `|`, found `(`
  --> src/lib.rs:26:22
   |
26 |     ..params::default()
   |                      ^ expected one of `::`, `:`, or `|`

error[E0425]: cannot find value `polygons` in this scope
  --> src/lib.rs:29:22
   |
29 |     println!("{:?}", polygons[0]);
   |                      ^^^^^^^^ not found in this scope

error[E0425]: cannot find value `polygons` in this scope
  --> src/lib.rs:29:22
   |
29 |     println!("{:?}", polygons[0]);
   |                      ^^^^^^^^ not found in this scope

error[E0658]: half-open range patterns are unstable
  --> src/lib.rs:26:5
   |
26 |     ..params::default()
   |     ^^^^^^^^^^^^^^^^^
   |
   = note: see issue #67264 <https://github.com/rust-lang/rust/issues/67264> for more information

error[E0658]: half-open range patterns are unstable
  --> src/lib.rs:26:5
   |
26 |     ..params::default()
   |     ^^^^^^^^^^^^^^^^^
   |
   = note: see issue #67264 <https://github.com/rust-lang/rust/issues/67264> for more information

error[E0308]: mismatched types
  --> src/lib.rs:16:43
   |
16 |         Params { n_circles: 5, distances: 10}
   |                                           ^^ expected struct `Vec`, found integer
   |
   = note: expected struct `Vec<f64>`
                found type `{integer}`

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0425, E0658.
For more information about an error, try `rustc --explain E0308`.
error[E0308]: mismatched types
  --> src/lib.rs:16:43
   |
16 |         Params { n_circles: 5, distances: 10}
   |                                           ^^ expected struct `Vec`, found integer
   |
   = note: expected struct `Vec<f64>`
                found type `{integer}`

error: could not compile `zonebuilder`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error[E0061]: this function takes 5 arguments but 4 arguments were supplied
  --> src/lib.rs:60:28
   |
60 |         let polygon_list = clockboard(Point::new(0.0, 0.0), 2, vec![1.0], None);
   |                            ^^^^^^^^^^ --------------------  -  ---------  ---- supplied 4 arguments
   |                            |
   |                            expected 5 arguments
   |
note: function defined here
  --> src/lib.rs:21:8
   |
21 | pub fn clockboard(
   |        ^^^^^^^^^^
22 |     centerpoint: Point<f64>,
   |     -----------------------
23 |     num_segments: usize,
   |     -------------------
24 |     radii: Vec<f64>,
   |     ---------------
25 |     boundary: Option<Polygon<f64>>,
   |     ------------------------------
26 |     ..params::default()
   |               ---------

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0061, E0308, E0425, E0658.
For more information about an error, try `rustc --explain E0061`.
error: build failed
```

File: src/lib.rs [hunk 1]
```diff
@@ -1,16 +1,35 @@
...
+    n_circles: usize,
+    distances: Vec<f64>,
+}
+
+impl Default for Params {
+    fn default() -> Self {
+        // todo: distances should be:
+        // zonebuilder::zb_100_triangular_numbers
+        // 1    3    6   10   15   21   28   36   45   55   66 ...
+        Params { n_circles: 5, distances: 10}
```
Comment (dabreegster) [inline src/lib.rs]: `distances` is a `Vec<f64>`, but here you've passed in a single f64. You can say `distances: vec![1.0, 3.0, 6.0, 10.0, 15.0]` instead. `vec![...]` is a macro that creates a vector and does `vector.push(x)` a bunch of times for you

File: src/lib.rs [hunk 2]
```diff
@@ -1,16 +1,35 @@
...
+    }
+}
+
 
 pub fn clockboard(
     centerpoint: Point<f64>,
     num_segments: usize,
     radii: Vec<f64>,
     boundary: Option<Polygon<f64>>,
+    ..params::default()
```
Comment (dabreegster) [inline src/lib.rs]: This `..` syntax is used when you create `Params`, not here. You should follow the syntax of the rest of the arguments -- `params: Params`

When you call `clockboard`, you'll have to pass in `Params`. That's when you can do the `let options = SomeOptions { foo: 42, ..Default::default() };` thing

File: src/lib.rs [hunk 3]
```diff
@@ -2,28 +2,37 @@ use geo::{LineString, Point, Polygon};
...
+//         // 1    3    6   10   15   21   28   36   45   55   66 ...
+//         Params { n_circles: 5, distances: 10.0}
+//     }
+// }
+
+// Trying again
+// https://doc.rust-lang.org/std/default/trait.Default.html
+#[derive(Default)]
+struct SomeOptions {
+    n_circles: usize 5,
```
Comment (dabreegster) [inline src/lib.rs]: This isn't the syntax. Check out https://doc.rust-lang.org/rust-by-example/custom_types/structs.html

**Comment by Robinlovelace** [comment]
Thanks for the feedback @dabreegster I think I'm making progress. Slowly...

**Comment by Robinlovelace** [comment]
Slight improvement, I get this:

```
cargo test
   Compiling zonebuilder v0.1.0 (/mnt/57982e2a-2874-4246-a6fe-115c199bc6bd/orgs/zonebuilders/zonebuilder-rust)
error[E0425]: cannot find value `n_circles` in this scope
  --> src/lib.rs:32:22
   |
32 |     println!("{:?}", n_circles);
   |                      ^^^^^^^^^ not found in this scope

```

File: src/lib.rs (hunk 3 above)
Comment (Robinlovelace) [reply src/lib.rs]: Updated. Now reads:

```
#[derive(Debug)]
pub struct Params {
    n_circles: usize,
    num_segments: usize,
    distances: Vec<f64>,
}

// https://doc.rust-lang.org/std/default/trait.Default.html

impl Default for Params {
    fn default() -> Self {
        // todo: distances should be:
        // zonebuilder::zb_100_triangular_numbers
        // 1    3    6   10   15   21   28   36   45   55   66 ...
        Params { n_circles: 5, num_segments: 12, distances: vec![1.0, 3.0, 6.0, 10.0, 15.0]}
    }
}
```

File: src/lib.rs [hunk 4]
```diff
@@ -1,16 +1,39 @@
...
 pub fn clockboard(
     centerpoint: Point<f64>,
     num_segments: usize,
     radii: Vec<f64>,
     boundary: Option<Polygon<f64>>,
+    params: Params,
 ) -> Vec<Polygon<f64>> {
+    // test options worked // no idea if this next line is useful (probably not)
+    // let options = Params {n_circles: 42 ..Default::Default} 
+    println!("{:?}", n_circles);
```
Comment (dabreegster) [inline src/lib.rs]: `params.n_circles` is how you access a field/member of a struct

Comment (Robinlovelace) [reply src/lib.rs]: Bingo! Figured that out just as you responded, thanks to autocomplete. On the samewavelength, just about.

```
    println!("{:?}", params.n_circles);
```

**Comment by Robinlovelace** [comment]
Final issue - looking into it:

```
error[E0061]: this function takes 5 arguments but 4 arguments were supplied
  --> src/lib.rs:64:28
   |
64 |         let polygon_list = clockboard(Point::new(0.0, 0.0), 2, vec![1.0], None);
   |                            ^^^^^^^^^^ --------------------  -  ---------  ---- supplied 4 arguments
   |                            |
   |                            expected 5 arguments

```

**Comment by Robinlovelace** [comment]
Heads-up @dabreegster here's the latest test output:

```
cargo test
   Compiling zonebuilder v0.1.0 (/mnt/57982e2a-2874-4246-a6fe-115c199bc6bd/orgs/zonebuilders/zonebuilder-rust)
warning: unused variable: `num_segments`
  --> src/lib.rs:26:5
   |
26 |     num_segments: usize,
   |     ^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_num_segments`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `boundary`
  --> src/lib.rs:28:5
   |
28 |     boundary: Option<Polygon<f64>>,
   |     ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_boundary`

warning: unused variable: `num_segments`
  --> src/lib.rs:26:5
   |
26 |     num_segments: usize,
   |     ^^^^^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_num_segments`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `boundary`
  --> src/lib.rs:28:5
   |
28 |     boundary: Option<Polygon<f64>>,
   |     ^^^^^^^^ help: if this is intentional, prefix it with an underscore: `_boundary`

warning: 2 warnings emitted

warning: 2 warnings emitted

    Finished test [unoptimized + debuginfo] target(s) in 0.58s
     Running target/debug/deps/zonebuilder-93165252b2a528ef

running 1 test
test tests::it_works ... FAILED

failures:

---- tests::it_works stdout ----
5
{"coordinates":[[[1.0,0.0],[0.998652088398823,0.05190381813189974],[0.9946119873266613,0.10366771274658874],[0.9878905881817251,0.15515213753592053],[0.9785060106677815,0.20621829959298818],[0.966483553946637,0.2567285335732631],[0.9518556284360696,0.3065466728160249],[0.9346616684370731,0.35553841642560485],[0.9149480258259531,0.40357169132285653],[0.8927678450978637,0.4505170082908268],[0.8681809200986439,0.4962478110547916],[0.8412535328311812,0.5406408174555976],[0.81205827477085,0.5835763517965671],[0.780673851171728,0.6249386674680185],[0.7471848688911463,0.6646162589796615],[0.7116816083045602,0.7025021625596853],[0.674259779925618,0.738494244510177],[0.6350202663875384,0.7724954765415133],[0.5940688504813663,0.8044141973434756],[0.5515159299842706,0.8341643596879366],[0.5074762200466557,0.8616657623969739],[0.4620684439403962,0.8868442665510675],[0.41541501300188644,0.9096319953545183],[0.36764169663272006,0.9299675171192889],[0.31887728324761866,0.9477960108739739],[0.26925323308363447,0.9630694141514463],[0.21890332380659488,0.9757465525567718],[0.16796328987016487,0.985793250766098],[0.11657045659975915,0.993182424657285],[0.06486336998774303,0.9978941543239108],[0.012981423197930875,0.9999157377758169],[-0.03893551921373762,0.9992417251814281],[-0.09074749834931424,0.995873933559535],[-0.14231483827328523,0.9898214418809327],[-0.1934985225541996,0.9811005665931204],[-0.2441605690283914,0.9697348176340438],[-0.2941644017754963,0.9557548350534589],[-0.3433752193029885,0.9391983064127749],[-0.3916603579471703,0.9201098651860519],[-0.43888964951095566,0.8985409704360455],[-0.48493577217431666,0.8745497680896721],[-0.5296745937313987,0.8482009341868693],[-0.5729855062290019,0.8195655005254274],[-0.6147517511043008,0.7887206631718202],[-0.654860733945285,0.7557495743542583],[-0.693204328025389,0.7207411182989831],[-0.72967916579403,0.6837896716141071],[-0.764186917537249,0.644994848866964],[-0.796634556457236,0.6044612340408464],[-0.8269346094561286,0.562298098595078],[-0.85500539294802,0.5186191068884773],[-0.8807712330634644,0.4735420097603428],[-0.9041626696528502,0.42718832709500737],[-0.9251166435386837,0.37968302022571404],[-0.9435766665119819,0.33115415506095647],[-0.9594929736144974,0.28173255684142967],[-0.9728226572962473,0.2315514574583112],[-0.9835297830866824,0.18074613628363823],[-0.9915854864676663,0.12945355548103554],[-0.9969680506871134,0.07781199077993989],[-0.9996629653035124,0.02596065870867831],[-0.9996629653035124,-0.02596065870867851],[-0.9969680506871134,-0.07781199077993964],[-0.9915854864676663,-0.1294535554810353],[-0.9835297830866824,-0.18074613628363798],[-0.9728226572962473,-0.2315514574583114],[-0.9594929736144974,-0.28173255684142984],[-0.943576666511982,-0.33115415506095625],[-0.9251166435386838,-0.3796830202257138],[-0.9041626696528502,-0.42718832709500715],[-0.8807712330634643,-0.47354200976034294],[-0.8550053929480199,-0.5186191068884776],[-0.8269346094561288,-0.5622980985950777],[-0.7966345564572361,-0.6044612340408462],[-0.7641869175372492,-0.6449948488669639],[-0.7296791657940299,-0.6837896716141072],[-0.6932043280253889,-0.7207411182989832],[-0.6548607339452852,-0.7557495743542582],[-0.614751751104301,-0.78872066317182],[-0.5729855062290025,-0.8195655005254271],[-0.5296745937313985,-0.8482009341868694],[-0.4849357721743169,-0.874549768089672],[-0.4388896495109555,-0.8985409704360456],[-0.3916603579471703,-0.9201098651860518],[-0.3433752193029889,-0.9391983064127747],[-0.2941644017754963,-0.9557548350534589],[-0.24416056902839162,-0.9697348176340437],[-0.19349852255419941,-0.9811005665931205],[-0.14231483827328523,-0.9898214418809327],[-0.09074749834931471,-0.995873933559535],[-0.03893551921373764,-0.9992417251814281],[0.012981423197930629,-0.9999157377758169],[0.06486336998774324,-0.9978941543239108],[0.11657045659975912,-0.993182424657285],[0.1679632898701644,-0.985793250766098],[0.21890332380659486,-0.9757465525567718],[0.26925323308363425,-0.9630694141514464],[0.3188772832476188,-0.9477960108739738],[0.36764169663272006,-0.9299675171192889],[0.41541501300188605,-0.9096319953545186],[0.4620684439403962,-0.8868442665510675],[0.5074762200466555,-0.861665762396974],[0.5515159299842709,-0.8341643596879363],[0.5940688504813662,-0.8044141973434757],[0.6350202663875381,-0.7724954765415136],[0.6742597799256181,-0.738494244510177],[0.7116816083045601,-0.7025021625596856],[0.7471848688911465,-0.6646162589796613],[0.780673851171728,-0.6249386674680185],[0.8120582747708497,-0.5835763517965675],[0.8412535328311812,-0.5406408174555974],[0.8681809200986438,-0.4962478110547918],[0.8927678450978638,-0.4505170082908266],[0.9149480258259531,-0.4035716913228566],[0.934661668437073,-0.35553841642560524],[0.9518556284360696,-0.30654667281602477],[0.9664835539466369,-0.2567285335732633],[0.9785060106677815,-0.2062182995929879],[0.9878905881817251,-0.1551521375359206],[0.9946119873266613,-0.10366771274658916],[0.998652088398823,-0.05190381813189963],[1.0,0.0]]],"type":"Polygon"}
thread 'tests::it_works' panicked at 'assertion failed: false', src/lib.rs:67:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::it_works

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

```

**Comment by dabreegster** [comment]
LGTM -- nice! You could also move some of the existing clockboard params into that struct

**Comment by dabreegster** [comment]
I would also just delete the `cfg(test)` stuff and do all of your working by `cargo run`, now that we have `main.rs`. We can add back the unit tests later and actually make it test stuff.

**Comment by Robinlovelace** [comment]
> I would also just delete the `cfg(test)` stuff and do all of your working by `cargo run`, now that we have `main.rs`. We can add back the unit tests later and actually make it test stuff.

Done in latest commit.

**Comment by Robinlovelace** [comment]
> You could remove `radii` and use `params.distances` now. And move `num_segments` and `boundary` into the `Params` struct. Basically wind up with `fn clockboard(center: Point<f64>, params: Params)`

WiP...

**Comment by Robinlovelace** [comment]
Yay, think it's working now. Pls take another look and comment/merge as appropriate @dabreegster. Many thanks!

**Comment by dabreegster** [comment]
Awesome, you're learning fast! Merging

**Pull request merged**